### ISSUE TITLE
SI WFE interpolation is suboptimal for SIs with few measurements from CV3

### ISSUE THREAD

**Issue by mperrin** [spacetelescope/webbpsf#246]
The interpolation of zernikes in WebbFieldDependentAberration uses `griddata` with cubic interpolation. This seems to work not very well for cases where we have few data points measured from the ISIM CV3 phase retrieval. The WFE model appears to "snap" from one to the other, rather than smoothly interpolating. Specific examples include the "right side" (more negative V2) of the MIRI FOV, and the "left side" of the NIRISS FOV, each of which only have 2 measurements in the corners. 

See attached where the PSF speckles are identical then change abruptly, rather than varying smoothly from place to place

![miri-psf-diff-example](https://user-images.githubusercontent.com/1151745/49470198-63407c80-f7d7-11e8-9a47-5db00dce378e.png)

![niriss-psf-diff-example](https://user-images.githubusercontent.com/1151745/49470310-a995db80-f7d7-11e8-89c2-56b8aa39c248.png)

Tried `method="linear"` but it did not work better. May need to investigate other methods of interpolation for these SIs.

**Comment by JarronL** [comment]
Is this because those sampled field points are outside the region where the SI WFE is sampled? In that case, griddata returns a NaN, so the code simply selects the closest field point.

**Comment by mperrin** [comment]
Hmm, maybe! That would make sense. Good suggestion for investigation, thanks.

**Comment by JarronL** [comment]
Now that I've verified the Zemax models have good agreement to the CV3 measurements, I can provide you with a WFE measurements for a border of field points just outside the NIRCam FOV. So, for NIRCam at least, we will always be sampling within a defined grid region.

**Comment by JarronL** [comment]
Here's some code that performs a linear extrapolation on regions outside the bounds of the measured field points. This would drop in around line 1115 of optics.py (inside `class WebbFieldDependentAberration(poppy.OpticalElement)`) to replace the line `cf = self.row[zkey]`. Rather than just picking the nearest neighbor, the code creates a fine-meshed cubic fit over the known field points, fixes any NaN's using `RegularGridInterpolator`, then again uses `RegularGridInterpolator` on the fixed data to extrapolate the requested field point. 

In principle, the first call of `RegularGridInterpolator` can be used to used to extrapolate the requested field point to eliminate the intermediate step, but I wanted to be sure to use all the real data rather than the trimmed data set.

```
from scipy.interpolate import griddata, RegularGridInterpolator

# Want to perform extrapolation if outside of bounds
if np.isnan(cf):
    #cf = self.row[zkey]

    # Create fine mesh grid
    dstep = 1. / 60. # 1" steps
    xgrid = np.arange(v2.min(), v2.max()+dstep, dstep)
    ygrid = np.arange(v3.min(), v3.max()+dstep, dstep)
    X, Y = np.meshgrid(xgrid,ygrid)

    # Cubic interpolation of all points
    zgrid = griddata((v2, v3), zvals, (X, Y), method='cubic')
    
    # There will be some NaNs along the border that need to be replaced
    ind_nan = np.isnan(zgrid)
    # Remove rows/cols 1 by 1 until no NaNs
    zgrid2 = zgrid
    ygrid2 = ygrid
    xgrid2 = xgrid
    while np.isnan(zgrid2.sum()):
        zgrid = zgrid2[1:-1,1:-1]
        ygrid = ygrid2[1:-1]
        xgrid = xgrid2[1:-1]

    # Create regular grid interpolator function for extrapolation of NaN's
    func = RegularGridInterpolator((ygrid2,xgrid2), zgrid2, method='linear',
                                   bounds_error=False, fill_value=None)

    # Replace NaNs
    pts = np.array([Y[ind_nan], X[ind_nan]]).transpose()
    zgrid[ind_nan] = func(pts)

    # Recreate function without NaNs
    func = RegularGridInterpolator((ygrid,xgrid), zgrid, method='linear',
                                   bounds_error=False, fill_value=None)
    cf = func( (v3_tel, v2_tel) )
```

**Comment by mperrin** [comment]
Thanks @JarronL, that looks nice. I'm a little uncertain about the double setup of the interpolation, which as you say is probably not strictly necessary. But I expect the calls to RegularGridInterpolator are fast enough that it isn't really a concern if that gets ran twice.

**Comment by mperrin** [comment]
This was (partly) addressed a long while ago, in PR #283, but still could be improved further, particularly for MIRI.
